File: server/rag.ts
/**
 * RAG layer — ChromaDB (REST API) + @xenova/transformers embeddings
 *
 * Uses all-MiniLM-L6-v2 (25 MB, cached locally after first download)
 * to generate semantic embeddings. ChromaDB stores vectors on disk for
 * persistence — no re-indexing needed on subsequent server restarts.
 */

import { pipeline, type FeatureExtractionPipeline } from '@xenova/transformers';
import { FITNESS_KNOWLEDGE } from './fitnessKnowledge.js';
import { NUTRITION_KNOWLEDGE } from './nutritionKnowledge.js';

// Bump version whenever knowledge base changes — forces re-index on next cold start
const COLLECTION_NAME = 'fitness_knowledge_v2';
const CHROMA_URL = process.env.CHROMA_URL ?? 'http://localhost:8000';
const CHROMA_API = `${CHROMA_URL}/api/v2/tenants/default_tenant/databases/default_database`;
const BATCH_SIZE = 50;

let collectionId: string | null = null;
let extractor: FeatureExtractionPipeline | null = null;
let totalDocs = 0;

// ── Embedding ────────────────────────────────────────────────────────

async function getExtractor(): Promise<FeatureExtractionPipeline> {
  if (!extractor) {
    console.log('Loading all-MiniLM-L6-v2 embedding model (downloads once, ~25 MB)...');
    extractor = await pipeline('feature-extraction', 'Xenova/all-MiniLM-L6-v2');
    console.log('✓ Embedding model ready');
  }
  return extractor;
}

async function embed(texts: string[]): Promise<number[][]> {
  const model = await getExtractor();
  const embeddings: number[][] = [];
  for (const text of texts) {
    const output = await model(text.slice(0, 512), { pooling: 'mean', normalize: true });
    embeddings.push(Array.from(output.data as Float32Array));
  }
  return embeddings;
}

// ── ChromaDB REST helpers ────────────────────────────────────────────

async function chromaFetch(path: string, options?: RequestInit) {
  const res = await fetch(`${CHROMA_API}${path}`, {
    headers: { 'Content-Type': 'application/json' },
    ...options,
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`ChromaDB ${options?.method ?? 'GET'} ${path} → ${res.status}: ${text}`);
  }
  return res.json();
}

async function getOrCreateCollection(): Promise<string> {
  // List existing collections
  const collections: { id: string; name: string }[] = await chromaFetch('/collections');
  const existing = collections.find(c => c.name === COLLECTION_NAME);
  if (existing) return existing.id;

  // Create new
  const created = await chromaFetch('/collections', {
    method: 'POST',
    body: JSON.stringify({ name: COLLECTION_NAME }),
  });
  return created.id;
}

async function countDocs(id: string): Promise<number> {
  const result = await chromaFetch(`/collections/${id}/count`);
  return typeof result === 'number' ? result : 0;
}

async function addBatch(
  id: string,
  batchIds: string[],
  batchEmbeddings: number[][],
  batchDocuments: string[],
  batchMetadatas: Record<string, string>[]
) {
  await chromaFetch(`/collections/${id}/add`, {
    method: 'POST',
    body: JSON.stringify({
      ids: batchIds,
      embeddings: batchEmbeddings,
      documents: batchDocuments,
      metadatas: batchMetadatas,
    }),
  });
}

// ── Build index ──────────────────────────────────────────────────────

export async function buildIndex(): Promise<void> {
  // Verify ChromaDB is reachable
  await fetch(`${CHROMA_URL}/api/v2/heartbeat`).then(r => {
    if (!r.ok) throw new Error('ChromaDB not reachable');
  });
  console.log('✓ Connected to ChromaDB at', CHROMA_URL);

  collectionId = await getOrCreateCollection();
  const existing = await countDocs(collectionId);

  if (existing > 0) {
    totalDocs = existing;
    console.log(`✓ ChromaDB: using persisted collection (${existing} docs) — skipping re-index`);
    return;
  }

  // Pre-load embedding model before fetching data
  await getExtractor();

  // ── Assemble documents ─────────────────────────────────────────────
  const ids: string[] = [];
  const texts: string[] = [];
  const metadatas: Record<string, string>[] = [];

  for (const k of FITNESS_KNOWLEDGE) {
    ids.push(k.id);
    texts.push(`${k.title}\n\n${k.content}`);
    metadatas.push({ type: 'knowledge', title: k.title, tags: k.tags.join(',') });
  }

  for (const n of NUTRITION_KNOWLEDGE) {
    ids.push(`nutrition-${n.id}`);
    texts.push(`${n.title}\n\n${n.content}`);
    metadatas.push({ type: 'nutrition', title: n.title, tags: n.tags.join(',') });
  }
  console.log(`✓ Prepared ${NUTRITION_KNOWLEDGE.length} nutrition knowledge chunks`);

  let exerciseCount = 0;
  try {
    console.log('Fetching exercise database from GitHub...');
    const res = await fetch(
      'https://raw.githubusercontent.com/yuhonas/free-exercise-db/main/dist/exercises.json'
    );
    if (res.ok) {
      const exercises: any[] = await res.json();
      for (const ex of exercises) {
        ids.push(`exercise-${ex.id}`);
        texts.push([
          `Exercise: ${ex.name}`,
          `Category: ${ex.category} | Level: ${ex.level} | Equipment: ${ex.equipment}`,
          `Primary muscles: ${(ex.primaryMuscles ?? []).join(', ')}`,
          (ex.secondaryMuscles ?? []).length ? `Secondary muscles: ${ex.secondaryMuscles.join(', ')}` : '',
          ex.force ? `Force: ${ex.force}` : '',
          ex.mechanic ? `Mechanic: ${ex.mechanic}` : '',
          `Instructions: ${(ex.instructions ?? []).join(' ')}`,
        ].filter(Boolean).join('\n'));
        metadatas.push({
          type: 'exercise',
          title: ex.name,
          tags: [...(ex.primaryMuscles ?? []), ...(ex.secondaryMuscles ?? []), ex.category, ex.level, ex.equipment].join(','),
        });
        exerciseCount++;
      }
      console.log(`✓ Prepared ${exerciseCount} exercises`);
    }
  } catch (err) {
    console.warn('Could not fetch exercises:', err);
  }

  // ── Embed + index in batches ───────────────────────────────────────
  console.log(`Embedding and indexing ${ids.length} documents (runs once)...`);
  for (let i = 0; i < ids.length; i += BATCH_SIZE) {
    const slice = (arr: any[]) => arr.slice(i, i + BATCH_SIZE);
    const batchTexts = slice(texts);
    const embeddings = await embed(batchTexts);
    await addBatch(collectionId!, slice(ids), embeddings, batchTexts, slice(metadatas));
    process.stdout.write(`\r  Indexed ${Math.min(i + BATCH_SIZE, ids.length)}/${ids.length} docs`);
  }
  console.log('\n✓ ChromaDB index complete');

  totalDocs = await countDocs(collectionId!);
  console.log(`✓ ${totalDocs} vectors persisted to disk`);
}

// ── Retrieval ────────────────────────────────────────────────────────

export async function retrieve(query: string, k = 8): Promise<string[]> {
  if (!collectionId) throw new Error('ChromaDB not initialized');
  const [queryEmbedding] = await embed([query]);

  const result = await chromaFetch(`/collections/${collectionId}/query`, {
    method: 'POST',
    body: JSON.stringify({
      query_embeddings: [queryEmbedding],
      n_results: k,
    }),
  });

  return (result.documents?.[0] ?? []).filter(Boolean);
}

export function formatContext(docs: string[]): string {
  return docs.join('\n\n---\n\n');
}

export function getStats() {
  return { totalDocs, chromaUrl: CHROMA_URL, collectionId };
}

File: server/rag.py
import os
import httpx
from fastembed import TextEmbedding
from fitness_knowledge import FITNESS_KNOWLEDGE
from nutrition_knowledge import NUTRITION_KNOWLEDGE

CHROMA_URL = os.getenv("CHROMA_URL", "http://chromadb.railway.internal:8000")
COLLECTION_NAME = "fitness_knowledge_v2"

_embedder: TextEmbedding | None = None
_collection_id: str | None = None

ALL_DOCS = FITNESS_KNOWLEDGE + NUTRITION_KNOWLEDGE


def _get_embedder() -> TextEmbedding:
    global _embedder
    if _embedder is None:
        _embedder = TextEmbedding("sentence-transformers/all-MiniLM-L6-v2")
    return _embedder


def _embed(texts: list[str]) -> list[list[float]]:
    embedder = _get_embedder()
    return [v.tolist() for v in embedder.embed(texts)]


async def _get_or_create_collection() -> str:
    global _collection_id
    if _collection_id:
        return _collection_id

    async with httpx.AsyncClient(timeout=30) as client:
        # Try to get existing collection
        resp = await client.get(f"{CHROMA_URL}/api/v1/collections/{COLLECTION_NAME}")
        if resp.status_code == 200:
            _collection_id = resp.json()["id"]
            print(f"✓ Using existing ChromaDB collection: {COLLECTION_NAME}")
            return _collection_id

        # Create new collection
        resp = await client.post(
            f"{CHROMA_URL}/api/v1/collections",
            json={"name": COLLECTION_NAME, "metadata": {"hnsw:space": "cosine"}},
        )
        resp.raise_for_status()
        _collection_id = resp.json()["id"]
        print(f"✓ Created ChromaDB collection: {COLLECTION_NAME}")
        return _collection_id


async def build_index() -> None:
    collection_id = await _get_or_create_collection()

    async with httpx.AsyncClient(timeout=60) as client:
        # Check how many docs already exist
        resp = await client.get(f"{CHROMA_URL}/api/v1/collections/{collection_id}/count")
        count = resp.json() if resp.status_code == 200 else 0

        if isinstance(count, int) and count >= len(ALL_DOCS):
            print(f"✓ ChromaDB index already populated ({count} docs)")
            return

        print(f"  Indexing {len(ALL_DOCS)} documents into ChromaDB...")

        # Build all doc chunks (expand each doc with its tags concatenated)
        ids = []
        documents = []
        metadatas = []

        for doc in ALL_DOCS:
            text = f"{doc['title']}. Tags: {', '.join(doc['tags'])}. {doc['content']}"
            ids.append(doc["id"])
            documents.append(text)
            metadatas.append({"title": doc["title"], "tags": ", ".join(doc["tags"])})

        embeddings = _embed(documents)

        # Upsert in batches of 50
        batch_size = 50
        for i in range(0, len(ids), batch_size):
            batch_ids = ids[i : i + batch_size]
            batch_docs = documents[i : i + batch_size]
            batch_meta = metadatas[i : i + batch_size]
            batch_emb = embeddings[i : i + batch_size]

            resp = await client.post(
                f"{CHROMA_URL}/api/v1/collections/{collection_id}/upsert",
                json={
                    "ids": batch_ids,
                    "documents": batch_docs,
                    "metadatas": batch_meta,
                    "embeddings": batch_emb,
                },
            )
            resp.raise_for_status()

        print(f"✓ Indexed {len(ids)} documents into ChromaDB")


async def retrieve(query: str, n_results: int = 8) -> list[dict]:
    collection_id = await _get_or_create_collection()
    query_embedding = _embed([query])[0]

    async with httpx.AsyncClient(timeout=30) as client:
        resp = await client.post(
            f"{CHROMA_URL}/api/v1/collections/{collection_id}/query",
            json={
                "query_embeddings": [query_embedding],
                "n_results": min(n_results, len(ALL_DOCS)),
                "include": ["documents", "metadatas", "distances"],
            },
        )
        if resp.status_code != 200:
            return []

        data = resp.json()
        results = []
        docs = data.get("documents", [[]])[0]
        metas = data.get("metadatas", [[]])[0]
        distances = data.get("distances", [[]])[0]

        for doc, meta, dist in zip(docs, metas, distances):
            results.append({
                "content": doc,
                "title": meta.get("title", ""),
                "score": 1 - dist,  # cosine similarity
            })

        return results


def format_context(docs: list[dict]) -> str:
    if not docs:
        return ""
    parts = []
    for i, doc in enumerate(docs, 1):
        parts.append(f"[{i}] {doc['title']}\n{doc['content']}")
    return "\n\n".join(parts)


def get_stats() -> dict:
    return {"totalDocs": len(ALL_DOCS), "collection": COLLECTION_NAME}
